packages/protobuf/src/transform/index.ts

```diff
@@ -0,0 +1,933 @@
+      reportDiagnostic(program, {
+        code: "unsupported-field-type",
+        messageId: "recursive-map",
+        target: valueType,
+      });
+      return unreachable("recursive map");
+    }
+
+    // This is a core compile error.
+    if (!keyType || !valueType) return unreachable("nonexistent map key or value type");
+
+    // Key constraint (integral | string) is enforced by the type constraint on the `Map<>` type.
+    const keyProto = addType(keyType, relativeSource);
+    const valueProto = addType(valueType, relativeSource) as ProtoRef | ProtoScalar;
+
+    return map(keyProto[1] as "string" | ScalarIntegralName, valueProto);
+  }
+
+  function arrayToProto(t: Model, relativeSource: Model | Operation): ProtoType {
+    const valueType = (t as Model).templateMapper!.args[0];
+
+    // Nested arrays are not supported.
+    if (isArray(valueType)) {
+      reportDiagnostic(program, {
+        code: "nested-array",
+        target: valueType,
+      });
+      return ref("<unreachable>");
+    }
+
+    return addType(valueType, relativeSource);
+  }
+
+  function getProtoScalarsMap(program: Program): Map<Type, ProtoScalar> {
+    // The type references are different object identities in different programs, so we need to cache the map per program.
+    // This really only affects tests in our current use case, but someone could be using the compiler API to compile
+    // multiple programs and it also affects that.
+    let scalarMap;
+    if (_protoScalarsMap.has(program)) {
+      scalarMap = _protoScalarsMap.get(program)!;
+    } else {
+      scalarMap = new Map<Type, ProtoScalar>(
+        (
+          [
+            [program.resolveTypeReference("TypeSpec.bytes"), scalar("bytes")],
+            [program.resolveTypeReference("TypeSpec.boolean"), scalar("bool")],
+            [program.resolveTypeReference("TypeSpec.string"), scalar("string")],
+            [program.resolveTypeReference("TypeSpec.int32"), scalar("int32")],
+            [program.resolveTypeReference("TypeSpec.int64"), scalar("int64")],
+            [program.resolveTypeReference("TypeSpec.uint32"), scalar("uint32")],
+            [program.resolveTypeReference("TypeSpec.uint64"), scalar("uint64")],
+            [program.resolveTypeReference("TypeSpec.float32"), scalar("float")],
+            [program.resolveTypeReference("TypeSpec.float64"), scalar("double")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sfixed32"), scalar("sfixed32")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sfixed64"), scalar("sfixed64")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sint32"), scalar("sint32")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.sint64"), scalar("sint64")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.fixed32"), scalar("fixed32")],
+            [program.resolveTypeReference("TypeSpec.Protobuf.fixed64"), scalar("fixed64")],
+          ] as [[Type, unknown], ProtoScalar][]
```

Comment:
```suggestion
          ] satisfies [[Type, unknown], ProtoScalar][]
```

does `satisfies` work here instead?

Reply:
No. I tried to find a more elegant solution here, but it's difficult without many assertions. The core issue is that `resolveTypeReference` produces `Type | undefined`, but we know it will actually produce a `Type`.

In the latest commit I've made this a bit more resilient. It will validate that all the scalars resolved to something and throw an internal error if not.